### PR TITLE
Add Radon transform function to ximgproc

### PR DESCRIPTION
**Pull request by kuloPo**
This function calculates the Hough Space of a given image in any range.
Hough Transform is a discrete implementation of Radon Transform.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by AleksandrPanov** [comment]
I compared the current version of Radon Transform with the [scikit-image radon version](https://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.radon) based on [scikit-image example.](https://scikit-image.org/docs/dev/auto_examples/transform/plot_radon_transform.html)

- The current version always crops the image when it is rotated.
- The current version gives a slightly different result compared to the scikit-image version (this may be due to different interpolation or different rounding.). Need add more tests to check the meaning of the result (for example tests from radon.py.txt).
- It might be helpful to add CV_32FC1 for src/dst.

[radon.py.txt](https://github.com/opencv/opencv_contrib/files/7455701/radon.py.txt)
radon

**Comment by kuloPo** [comment]
Updates:

- Change function name to RadonTransform
- Fix always cropping corner when rotate
- transpose the output - now each column is the integral
- add support for float number

The small difference may caused by:
- rounding from rad to deg in scikit
- different rotation method
- copying image to a bigger background to capture the rotated image

**Comment by AleksandrPanov** [comment]
Thank you, I think it got better after your update
radon2
:

**Comment by AleksandrPanov** [comment]
LGTM + minor comments:
- this patch should go into 4.x branch (only bug fixes get into 3.4 branch).
- rebase your commits from 3.4 onto 4.x branch. For example:
   	`git rebase -i --onto upstream/3.4 upstream/master`
   	(check list of your commits, save and quit (Esc + "wq" + Enter)
   	where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).

- push rebased commits into source branch of your fork (with `--force` option)

Note: no need to re-open PR, apply changes "inplace".

**Comment by kuloPo** [comment]
The branch is rebased to 4.x

File: modules/ximgproc/perf/perf_radon_transform.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "perf_precomp.hpp"
+
+namespace opencv_test {
+    namespace {
+
+        typedef tuple<Size, MatType> RadonTransformPerfTestParam;
```
Comment (alalek) [inline modules/ximgproc/perf/perf_radon_transform.cpp:10]: Please avoid indentation in namespaces

File: modules/ximgproc/samples/radon_transform_demo.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <opencv2/highgui.hpp>
+#include <opencv2/ximgproc/radon_transform.hpp>
+
+int main() {
+	cv::Mat src = cv::imread("peilin_plane.png", cv::IMREAD_GRAYSCALE);
+	cv::Mat radon;
```
Comment (alalek) [inline modules/ximgproc/samples/radon_transform_demo.cpp:10]: Please use 4 spaces for indentation (no tabs)

File: modules/ximgproc/samples/radon_transform_demo.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include <opencv2/highgui.hpp>
+#include <opencv2/ximgproc/radon_transform.hpp>
+
+int main() {
+	cv::Mat src = cv::imread("peilin_plane.png", cv::IMREAD_GRAYSCALE);
```
Comment (alalek) [inline modules/ximgproc/samples/radon_transform_demo.cpp:9]: > cv::

In OpenCV samples we add `using namespace cv;` after includes and removing all `cv::` prefixes.

File: modules/ximgproc/test/test_radon_transform.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+namespace opencv_test {namespace {
+
+TEST(RadonTransformTest, output_size)
+{
+    Mat src(Size(256, 256), CV_8U, Scalar(0));
+    circle(src, Point(128, 128), 64, Scalar(255), FILLED);
+    Mat radon;
+    cv::ximgproc::RadonTransform(src, radon);
+
+    ASSERT_EQ(radon.rows, 363);
```
Comment (alalek) [inline modules/ximgproc/test/test_radon_transform.cpp:16]: > _EQ

uses different order of parameters: `(expected, actual)`.

Signature for `_EQ` is `ASSERT_EQ(expected_reference, actual_result);`.
Correct order of parameters are required to emit valid error messages.

Reference: https://github.com/opencv/opencv/blob/4.0.0/modules/ts/include/opencv2/ts/ts_gtest.h#L8196-L8200

```
GTEST_API_ AssertionResult EqFailure(const char* expected_expression,
                                     const char* actual_expression,
                                     const std::string& expected_value,
                                     const std::string& actual_value,
                                     bool ignoring_case);
```

File: modules/ximgproc/perf/perf_radon_transform.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            Size srcSize = get<0>(GetParam());
+            int  srcType = get<1>(GetParam());
+
+            Mat src(srcSize, srcType);
+            Mat radon;
+
+            declare.in(src, WARMUP_RNG);
+
+            TEST_CYCLE_N(3)
```
Comment (alalek) [inline modules/ximgproc/perf/perf_radon_transform.cpp:28]: > TEST_CYCLE_N

is misused here. No need to specify amount of inner samples here.

Please use:
- `TEST_CYCLE()`
- or `PERF_SAMPLE_BEGIN()` + `PERF_SAMPLE_END()`

File: modules/ximgproc/test/test_radon_transform.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+TEST(RadonTransformTest, accuracy_by_pixel)
+{
+    Mat src(Size(256, 256), CV_8U, Scalar(0));
+    circle(src, Point(128, 128), 64, Scalar(255), FILLED);
+    Mat radon;
+    cv::ximgproc::RadonTransform(src, radon);
+
+    ASSERT_EQ(radon.at<int>(0, 0), 0);
```
Comment (alalek) [inline modules/ximgproc/test/test_radon_transform.cpp:54]: Before using `.at<int>` it makes sense to validate type.

```
ASSERT_EQ(CV_32SC1, radon.type());

EXPECT_EQ(0, radon.at<int>(0, 0));
```

(prefer to use `EXPECT_` for non-critical failures - e.g. code below would not crash)

**Review by alalek — COMMENTED** [review]
@kuloPo Thank you for updates!

Please take a look on the comments below.

**Comment by kuloPo** [comment]
All the issues are fixed.

**Review by AleksandrPanov — APPROVED** [review]
👍

**Pull request merged**